File: python/features/steps/wiki_steps.py
"""Behave steps for wiki rendering and wiki-issue cross-linking."""

from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path

from behave import given, use_step_matcher, when, then

from kanbus.models import IssueComment

from features.steps.shared import (
    build_issue,
    load_project_directory,
    read_issue_file,
    run_cli,
    write_issue_file,
)


@given("3 open tasks exist")
def given_three_open_tasks(context: object) -> None:
    project_dir = load_project_directory(context)
    issues = [
        build_issue("kanbus-open01", "Open 1", "task", "open", None, []),
        build_issue("kanbus-open02", "Open 2", "task", "open", None, []),
        build_issue("kanbus-open03", "Open 3", "task", "open", None, []),
    ]
    for issue in issues:
        write_issue_file(project_dir, issue)


@given("3 open tasks and 2 closed tasks exist")
def given_open_and_closed_tasks(context: object) -> None:
    project_dir = load_project_directory(context)
    issues = [
        build_issue("kanbus-open01", "Open 1", "task", "open", None, []),
        build_issue("kanbus-open02", "Open 2", "task", "open", None, []),
        build_issue("kanbus-open03", "Open 3", "task", "open", None, []),
        build_issue("kanbus-closed01", "Closed 1", "task", "closed", None, []),
        build_issue("kanbus-closed02", "Closed 2", "task", "closed", None, []),
    ]
    for issue in issues:
        write_issue_file(project_dir, issue)


@given('open tasks "{first}" and "{second}" exist')
def given_open_tasks_alpha_beta(context: object, first: str, second: str) -> None:
    project_dir = load_project_directory(context)
    issues = [
        build_issue("kanbus-alpha", first, "task", "open", None, []),
        build_issue("kanbus-beta", second, "task", "open", None, []),
    ]
    for issue in issues:
        write_issue_file(project_dir, issue)


@given('open tasks "Urgent" and "Later" exist with priorities 1 and 3')
def given_open_tasks_with_priorities(context: object) -> None:
    project_dir = load_project_directory(context)
    urgent = build_issue("kanbus-urgent", "Urgent", "task", "open", None, [])
    later = build_issue("kanbus-later", "Later", "task", "open", None, [])
    urgent = urgent.model_copy(update={"priority": 1})
    later = later.model_copy(update={"priority": 3})
    for issue in [urgent, later]:
        write_issue_file(project_dir, issue)


@given('a wiki page "{filename}" with content "{content}"')
def given_wiki_page_with_content_string(
    context: object, filename: str, content: str
) -> None:
    project_dir = load_project_directory(context)
    wiki_subdir = getattr(context, "wiki_directory", "wiki")
    if wiki_subdir.startswith("../"):
        repo_root = Path(context.working_directory)
        wiki_dir = repo_root / wiki_subdir.lstrip("../").lstrip("..\\")
    else:
        wiki_dir = project_dir / wiki_subdir
    wiki_dir.mkdir(parents=True, exist_ok=True)
    target = wiki_dir / filename
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(content, encoding="utf-8")


@given('a wiki page "{filename}" with content')
@given('a wiki page "{filename}" with content:')
def given_wiki_page_with_content(context: object, filename: str) -> None:
    project_dir = load_project_directory(context)
    wiki_subdir = getattr(context, "wiki_directory", "wiki")
    if wiki_subdir.startswith("../"):
        repo_root = Path(context.working_directory)
        wiki_dir = repo_root / wiki_subdir.lstrip("../").lstrip("..\\")
    else:
        wiki_dir = project_dir / wiki_subdir
    wiki_dir.mkdir(parents=True, exist_ok=True)
    content = context.text or ""
    target = wiki_dir / filename
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(content, encoding="utf-8")


@given('a raw wiki page "{filename}" with content')
@given('a raw wiki page "{filename}" with content:')
def given_raw_wiki_page_with_content(context: object, filename: str) -> None:
    working_directory = Path(context.working_directory)
    content = context.text or ""
    (working_directory / filename).write_text(content, encoding="utf-8")


@when('I render the wiki page "{filename}" by absolute path')
def when_render_page_absolute(context: object, filename: str) -> None:
    project_dir = load_project_directory(context)
    page_path = (project_dir / "wiki" / filename).resolve()
    run_cli(context, f"kanbus wiki render {page_path}")


@then('"{first}" should appear before "{second}" in the output')
def then_first_before_second(context: object, first: str, second: str) -> None:
    stdout = context.result.stdout
    first_index = stdout.find(first)
    second_index = stdout.find(second)
    assert first_index != -1
    assert second_index != -1
    assert first_index < second_index


use_step_matcher("re")


@given(r'issue "(?P<identifier>[^"]+)" has description containing(?:\s*:)?')
def given_issue_has_description_containing(context: object, identifier: str) -> None:
    project_dir = load_project_directory(context)
    content = context.text or ""
    try:
        issue = read_issue_file(project_dir, identifier)
    except Exception:
        issue = build_issue(identifier, "Title", "task", "open", None, [])
    issue = issue.model_copy(update={"description": content})
    write_issue_file(project_dir, issue)


use_step_matcher("parse")


@given('a comment on issue "{identifier}" contains "{text}"')
def given_comment_contains(context: object, identifier: str, text: str) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    author = getattr(context, "current_user", "dev@example.com")
    comment = IssueComment(
        author=author,
        text=text,
        created_at=datetime.now(timezone.utc),
    )
    comments = list(issue.comments) + [comment]
    issue = issue.model_copy(update={"comments": comments})
    write_issue_file(project_dir, issue)


use_step_matcher("re")


@given(r'a comment on issue "(?P<identifier>[^"]+)" contains(?:\s*:)?')
def given_comment_contains_multiline(context: object, identifier: str) -> None:
    text = context.text or ""
    given_comment_contains(context, identifier, text)


use_step_matcher("parse")


@then('the rendered description should contain a link to wiki path "{path}"')
def then_rendered_description_has_wiki_link(context: object, path: str) -> None:
    stdout = context.result.stdout
    assert path in stdout, f"expected wiki path {path!r} in output, got: {stdout[:500]}"


@then('the rendered comments should contain a link to wiki path "{path}"')
def then_rendered_comments_has_wiki_link(context: object, path: str) -> None:
    stdout = context.result.stdout
    assert path in stdout, f"expected wiki path {path!r} in output, got: {stdout[:500]}"


@then('the rendered description should contain "{text}"')
def then_rendered_description_contains(context: object, text: str) -> None:
    stdout = context.result.stdout
    assert text in stdout, f"expected {text!r} in output, got: {stdout[:500]}"


@then('the rendered comments should contain "{text}"')
def then_rendered_comments_contains(context: object, text: str) -> None:
    stdout = context.result.stdout
    assert text in stdout, f"expected {text!r} in output, got: {stdout[:500]}"


@then('the rendered wiki should contain a link to issue "{identifier}"')
def then_rendered_wiki_has_issue_link(context: object, identifier: str) -> None:
    stdout = context.result.stdout
    assert (
        identifier in stdout
    ), f"expected issue {identifier!r} in output, got: {stdout[:500]}"


@given("a Kanbus project with AI configured")
def given_project_with_ai_configured(context: object) -> None:
    """Create a Kanbus project with AI config and test mock enabled."""
    import copy
    import os
    import yaml

    from features.steps.shared import initialize_default_project
    from kanbus.config import DEFAULT_CONFIGURATION

    initialize_default_project(context)
    repository = Path(context.working_directory)
    config_path = repository / ".kanbus.yml"
    payload = copy.deepcopy(DEFAULT_CONFIGURATION)
    payload["ai"] = {"provider": "openai", "model": "gpt-4o"}
    config_path.write_text(
        yaml.safe_dump(payload, sort_keys=False),
        encoding="utf-8",
    )
    context._ai_mock_env = os.environ.get("KANBUS_TEST_AI_MOCK")
    os.environ["KANBUS_TEST_AI_MOCK"] = "1"


def _ai_calls_log_path(context: object) -> Path:
    """Return path to ai_calls.log in project cache."""
    project_dir = load_project_directory(context)
    return project_dir / ".cache" / "ai_calls.log"


def _ai_call_count(context: object) -> int:
    path = _ai_calls_log_path(context)
    if not path.exists():
        return 0
    return len(path.read_text(encoding="utf-8").strip().splitlines())


@then("the AI provider API should be called")
def then_ai_provider_api_called(context: object) -> None:
    count = _ai_call_count(context)
    assert count >= 1, f"expected at least 1 API call, got {count}"
    context._ai_call_count_after_first_render = count


@then("the AI provider API should not be called")
def then_ai_provider_api_not_called(context: object) -> None:
    count_before = getattr(context, "_ai_call_count_after_first_render", 0)
    count_after = _ai_call_count(context)
    assert (
        count_after == count_before
    ), f"expected no new API calls (count was {count_before}), got {count_after}"


@then('the rendered wiki should contain a generated summary for "{identifier}"')
def then_rendered_wiki_has_generated_summary(context: object, identifier: str) -> None:
    """Verify rendered wiki output contains the mock generated summary."""
    stdout = context.result.stdout
    expected = f"Generated summary for {identifier}"
    assert expected in stdout, f"expected {expected!r} in output, got: {stdout[:500]}"


@given('a Kanbus project with wiki_directory set to "{value}"')
def given_project_with_wiki_directory(context: object, value: str) -> None:
    import copy
    import yaml

    from features.steps.shared import initialize_default_project
    from kanbus.config import DEFAULT_CONFIGURATION

    initialize_default_project(context)
    repository = Path(context.working_directory)
    config_path = repository / ".kanbus.yml"
    payload = copy.deepcopy(DEFAULT_CONFIGURATION)
    payload["wiki_directory"] = value
    config_path.write_text(
        yaml.safe_dump(payload, sort_keys=False),
        encoding="utf-8",
    )
    context.wiki_directory = value


def _wiki_render_cache_dir(context: object) -> Path:
    project_dir = load_project_directory(context)
    return project_dir / ".cache" / "wiki_render"


@then("a cached rendered file should exist")
def then_cached_rendered_file_exists(context: object) -> None:
    cache_dir = _wiki_render_cache_dir(context)
    assert cache_dir.exists(), f"expected cache dir {cache_dir} to exist"
    md_files = list(cache_dir.glob("*.md"))
    assert md_files, f"expected at least one cached .md file in {cache_dir}"


@then("the command should use the cache")
def then_command_uses_cache(context: object) -> None:
    project_dir = load_project_directory(context)
    log_path = project_dir / ".cache" / "wiki_cache_hits.log"
    assert log_path.exists(), f"expected cache hit log {log_path} to exist"
    lines = log_path.read_text(encoding="utf-8").strip().splitlines()
    assert len(lines) >= 1, f"expected at least 1 cache hit, got {len(lines)}"


@then('the wiki root should be "{expected}"')
def then_wiki_root_is(context: object, expected: str) -> None:
    repo = Path(context.working_directory)
    wiki_path = (repo / expected).resolve()
    assert wiki_path.exists(), f"expected wiki root {expected} to exist at {wiki_path}"


@given('the Kanbus configuration has wiki_directory set to "{value}"')
def given_config_wiki_directory(context: object, value: str) -> None:
    import yaml

    repository = Path(context.working_directory)
    config_path = repository / ".kanbus.yml"
    payload = yaml.safe_load(config_path.read_text(encoding="utf-8")) or {}
    payload["wiki_directory"] = value
    config_path.write_text(
        yaml.safe_dump(payload, sort_keys=False),
        encoding="utf-8",
    )

File: rust/features/steps/wiki_steps.rs
use std::fs;
use std::path::PathBuf;
use std::process::Command;

use chrono::{TimeZone, Utc};
use cucumber::{gherkin::Step, given, then, when};
use serde_yaml::{Mapping, Value};
use tempfile::TempDir;

use kanbus::cli::run_from_args_with_output;
use kanbus::file_io::load_project_directory;
use kanbus::models::{IssueComment, IssueData};

use crate::step_definitions::initialization_steps::KanbusWorld;

fn run_cli(world: &mut KanbusWorld, command: &str) {
    let args = shell_words::split(command).expect("parse command");
    let cwd = world
        .working_directory
        .as_ref()
        .expect("working directory not set");

    match run_from_args_with_output(args, cwd.as_path()) {
        Ok(output) => {
            world.exit_code = Some(0);
            world.stdout = Some(output.stdout);
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

fn load_project_dir(world: &KanbusWorld) -> PathBuf {
    let cwd = world.working_directory.as_ref().expect("cwd");
    load_project_directory(cwd).expect("project dir")
}

fn write_issue_file(project_dir: &PathBuf, issue: &IssueData) {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{}.json", issue.identifier));
    let contents = serde_json::to_string_pretty(issue).expect("serialize issue");
    fs::write(issue_path, contents).expect("write issue");
}

fn build_issue(identifier: &str, title: &str, status: &str) -> IssueData {
    let timestamp = Utc.with_ymd_and_hms(2026, 2, 11, 0, 0, 0).unwrap();
    IssueData {
        identifier: identifier.to_string(),
        title: title.to_string(),
        description: "".to_string(),
        issue_type: "task".to_string(),
        status: status.to_string(),
        priority: 2,
        assignee: None,
        creator: None,
        parent: None,
        labels: Vec::new(),
        dependencies: Vec::new(),
        comments: Vec::new(),
        created_at: timestamp,
        updated_at: timestamp,
        closed_at: None,
        custom: std::collections::BTreeMap::new(),
    }
}

#[given("3 open tasks exist")]
fn given_three_open_tasks(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issues = vec![
        build_issue("kanbus-open01", "Open 1", "open"),
        build_issue("kanbus-open02", "Open 2", "open"),
        build_issue("kanbus-open03", "Open 3", "open"),
    ];
    for issue in issues {
        write_issue_file(&project_dir, &issue);
    }
}

#[given("3 open tasks and 2 closed tasks exist")]
fn given_open_and_closed_tasks(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issues = vec![
        build_issue("kanbus-open01", "Open 1", "open"),
        build_issue("kanbus-open02", "Open 2", "open"),
        build_issue("kanbus-open03", "Open 3", "open"),
        build_issue("kanbus-closed01", "Closed 1", "closed"),
        build_issue("kanbus-closed02", "Closed 2", "closed"),
    ];
    for issue in issues {
        write_issue_file(&project_dir, &issue);
    }
}

#[given(expr = "open tasks {string} and {string} exist")]
fn given_open_tasks(world: &mut KanbusWorld, first: String, second: String) {
    let project_dir = load_project_dir(world);
    let issues = vec![
        build_issue("kanbus-alpha", &first, "open"),
        build_issue("kanbus-beta", &second, "open"),
    ];
    for issue in issues {
        write_issue_file(&project_dir, &issue);
    }
}

#[given("open tasks \"Urgent\" and \"Later\" exist with priorities 1 and 3")]
fn given_open_tasks_with_priorities(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let mut urgent = build_issue("kanbus-urgent", "Urgent", "open");
    let mut later = build_issue("kanbus-later", "Later", "open");
    urgent.priority = 1;
    later.priority = 3;
    for issue in vec![urgent, later] {
        write_issue_file(&project_dir, &issue);
    }
}

#[given(expr = "a wiki page {string} with content {string}")]
fn given_wiki_page_with_content_string(world: &mut KanbusWorld, filename: String, content: String) {
    let project_dir = load_project_dir(world);
    let wiki_subdir = world
        .wiki_directory
        .as_ref()
        .map(|s| s.as_str())
        .unwrap_or("wiki");
    let wiki_dir = if wiki_subdir.starts_with("../") {
        let cwd = world.working_directory.as_ref().expect("working dir");
        cwd.join(
            wiki_subdir
                .trim_start_matches("../")
                .trim_start_matches("..\\"),
        )
    } else {
        project_dir.join(wiki_subdir)
    };
    fs::create_dir_all(&wiki_dir).expect("create wiki dir");
    let target = wiki_dir.join(&filename);
    if let Some(parent) = target.parent() {
        fs::create_dir_all(parent).expect("create wiki parent dir");
    }
    fs::write(target, content).expect("write wiki page");
}

#[given(expr = "a wiki page {string} with content")]
#[given(expr = "a wiki page {string} with content:")]
fn given_wiki_page_with_content(world: &mut KanbusWorld, filename: String, step: &Step) {
    let project_dir = load_project_dir(world);
    let wiki_subdir = world
        .wiki_directory
        .as_ref()
        .map(|s| s.as_str())
        .unwrap_or("wiki");
    let wiki_dir = if wiki_subdir.starts_with("../") {
        let cwd = world.working_directory.as_ref().expect("working dir");
        cwd.join(
            wiki_subdir
                .trim_start_matches("../")
                .trim_start_matches("..\\"),
        )
    } else {
        project_dir.join(wiki_subdir)
    };
    fs::create_dir_all(&wiki_dir).expect("create wiki dir");
    let content = step.docstring().map(|s| s.as_str()).unwrap_or("");
    let target = wiki_dir.join(&filename);
    if let Some(parent) = target.parent() {
        fs::create_dir_all(parent).expect("create wiki parent dir");
    }
    fs::write(target, content).expect("write wiki page");
}

#[given(expr = "a raw wiki page {string} with content")]
#[given(expr = "a raw wiki page {string} with content:")]
fn given_raw_wiki_page_with_content(world: &mut KanbusWorld, filename: String, step: &Step) {
    let cwd = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let content = step.docstring().expect("content not found");
    fs::write(cwd.join(filename), content).expect("write wiki page");
}

#[when(expr = "I render the wiki page {string} by absolute path")]
fn when_render_absolute(world: &mut KanbusWorld, filename: String) {
    let project_dir = load_project_dir(world);
    let page_path = project_dir.join("wiki").join(filename);
    run_cli(
        world,
        &format!("kanbus wiki render {}", page_path.display()),
    );
}

#[then(expr = "{string} should appear before {string} in the output")]
fn then_text_before_text(world: &mut KanbusWorld, first: String, second: String) {
    let stdout = world.stdout.as_ref().expect("stdout");
    let first_index = stdout.find(&first).expect("first value in stdout");
    let second_index = stdout.find(&second).expect("second value in stdout");
    assert!(first_index < second_index);
}

fn read_issue_file(project_dir: &PathBuf, identifier: &str) -> IssueData {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    let contents = fs::read_to_string(&issue_path).expect("read issue");
    serde_json::from_str(&contents).expect("parse issue")
}

#[given(expr = "issue {string} has description containing:")]
fn given_issue_has_description_containing(
    world: &mut KanbusWorld,
    identifier: String,
    step: &Step,
) {
    let project_dir = load_project_dir(world);
    let content = step.docstring().map(|s| s.as_str()).unwrap_or("");
    let mut issue = if project_dir
        .join("issues")
        .join(format!("{identifier}.json"))
        .exists()
    {
        read_issue_file(&project_dir, &identifier)
    } else {
        build_issue(&identifier, "Title", "open")
    };
    issue.description = content.to_string();
    write_issue_file(&project_dir, &issue);
}

#[given(expr = "a comment on issue {string} contains {string}")]
fn given_comment_contains(world: &mut KanbusWorld, identifier: String, text: String) {
    let project_dir = load_project_dir(world);
    let mut issue = read_issue_file(&project_dir, &identifier);
    let author = world.current_user.as_deref().unwrap_or("dev@example.com");
    issue.comments.push(IssueComment {
        id: None,
        author: author.to_string(),
        text,
        created_at: Utc::now(),
    });
    write_issue_file(&project_dir, &issue);
}

#[given(expr = "a comment on issue {string} contains:")]
fn given_comment_contains_multiline(world: &mut KanbusWorld, identifier: String, step: &Step) {
    let text = step
        .docstring()
        .map(|s| s.as_str())
        .unwrap_or("")
        .to_string();
    given_comment_contains(world, identifier, text);
}

#[then(expr = "the rendered description should contain a link to wiki path {string}")]
fn then_rendered_description_has_wiki_link(world: &mut KanbusWorld, path: String) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(
        stdout.contains(&path),
        "expected wiki path {path:?} in output"
    );
}

#[then(expr = "the rendered comments should contain a link to wiki path {string}")]
fn then_rendered_comments_has_wiki_link(world: &mut KanbusWorld, path: String) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(
        stdout.contains(&path),
        "expected wiki path {path:?} in output"
    );
}

#[then(expr = "the rendered description should contain {string}")]
fn then_rendered_description_contains(world: &mut KanbusWorld, text: String) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(stdout.contains(&text), "expected {text:?} in output");
}

#[then(expr = "the rendered comments should contain {string}")]
fn then_rendered_comments_contains(world: &mut KanbusWorld, text: String) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(stdout.contains(&text), "expected {text:?} in output");
}

#[then(expr = "the rendered wiki should contain a link to issue {string}")]
fn then_rendered_wiki_has_issue_link(world: &mut KanbusWorld, identifier: String) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(
        stdout.contains(&identifier),
        "expected issue {identifier:?} in output"
    );
}

#[given("a Kanbus project with AI configured")]
fn given_project_with_ai_configured(world: &mut KanbusWorld) {
    std::env::set_var("KANBUS_NO_DAEMON", "1");
    let temp_dir = TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join("repo");
    fs::create_dir_all(&repo_path).expect("create repo dir");
    Command::new("git")
        .args(["init"])
        .current_dir(&repo_path)
        .output()
        .expect("git init failed");
    world.working_directory = Some(repo_path.clone());
    world.temp_dir = Some(temp_dir);
    run_cli(world, "kanbus init");
    assert_eq!(world.exit_code, Some(0));

    let config_path = repo_path.join(".kanbus.yml");
    let contents = fs::read_to_string(&config_path).expect("read config");
    let mut mapping: Mapping = serde_yaml::from_str(&contents).expect("parse config");
    let mut ai_block = Mapping::new();
    ai_block.insert(
        Value::String("provider".to_string()),
        Value::String("openai".to_string()),
    );
    ai_block.insert(
        Value::String("model".to_string()),
        Value::String("gpt-4o".to_string()),
    );
    mapping.insert(Value::String("ai".to_string()), Value::Mapping(ai_block));
    let yaml = serde_yaml::to_string(&mapping).expect("serialize config");
    fs::write(config_path, yaml).expect("write config");

    world.jira_unset_env_vars.push((
        String::from("KANBUS_TEST_AI_MOCK"),
        std::env::var("KANBUS_TEST_AI_MOCK").ok(),
    ));
    std::env::set_var("KANBUS_TEST_AI_MOCK", "1");
}

fn ai_call_count(world: &KanbusWorld) -> usize {
    let project_dir = load_project_dir(world);
    let log_path = project_dir.join(".cache").join("ai_calls.log");
    if log_path.exists() {
        fs::read_to_string(&log_path)
            .unwrap_or_default()
            .lines()
            .filter(|l| !l.is_empty())
            .count()
    } else {
        0
    }
}

#[then("the AI provider API should be called")]
fn then_ai_provider_api_called(world: &mut KanbusWorld) {
    let count = ai_call_count(world);
    assert!(count >= 1, "expected at least 1 API call, got {}", count);
    world.ai_call_count_after_first_render = Some(count);
}

#[then("the AI provider API should not be called")]
fn then_ai_provider_api_not_called(world: &mut KanbusWorld) {
    let count = ai_call_count(world);
    let expected = world.ai_call_count_after_first_render.unwrap_or(1);
    assert_eq!(
        count, expected,
        "expected no new API calls (count should be {}), got {}",
        expected, count
    );
}

#[then("a cached rendered file should exist")]
fn then_cached_rendered_file_exists(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let cache_dir = project_dir.join(".cache").join("wiki_render");
    assert!(
        cache_dir.exists(),
        "expected cache dir {} to exist",
        cache_dir.display()
    );
    let md_files: Vec<_> = fs::read_dir(&cache_dir)
        .unwrap_or_else(|_| panic!("read dir {}", cache_dir.display()))
        .filter_map(|e| e.ok())
        .filter(|e| e.path().extension().map_or(false, |ext| ext == "md"))
        .collect();
    assert!(
        !md_files.is_empty(),
        "expected at least one cached .md file in {}",
        cache_dir.display()
    );
}

#[then("the command should use the cache")]
fn then_command_uses_cache(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let log_path = project_dir.join(".cache").join("wiki_cache_hits.log");
    assert!(
        log_path.exists(),
        "expected cache hit log {} to exist",
        log_path.display()
    );
    let content = fs::read_to_string(&log_path).unwrap_or_default();
    let lines: Vec<_> = content.lines().filter(|l| !l.is_empty()).collect();
    assert!(
        !lines.is_empty(),
        "expected at least 1 cache hit, got {}",
        lines.len()
    );
}

#[then(expr = "the rendered wiki should contain a generated summary for {string}")]
fn then_rendered_wiki_has_generated_summary(world: &mut KanbusWorld, identifier: String) {
    let stdout = world.stdout.as_ref().expect("stdout");
    let expected = format!("Generated summary for {}", identifier);
    assert!(
        stdout.contains(&expected),
        "expected {expected:?} in output"
    );
}

#[given(expr = "a Kanbus project with wiki_directory set to {string}")]
fn given_project_with_wiki_directory(world: &mut KanbusWorld, value: String) {
    std::env::set_var("KANBUS_NO_DAEMON", "1");
    let temp_dir = TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join("repo");
    fs::create_dir_all(&repo_path).expect("create repo dir");
    Command::new("git")
        .args(["init"])
        .current_dir(&repo_path)
        .output()
        .expect("git init failed");
    world.working_directory = Some(repo_path.clone());
    world.temp_dir = Some(temp_dir);
    run_cli(world, "kanbus init");
    assert_eq!(world.exit_code, Some(0));

    let config_path = repo_path.join(".kanbus.yml");
    let contents = fs::read_to_string(&config_path).expect("read config");
    let mut mapping: Mapping = serde_yaml::from_str(&contents).expect("parse config");
    mapping.insert(
        Value::String("wiki_directory".to_string()),
        Value::String(value.clone()),
    );
    let yaml = serde_yaml::to_string(&mapping).expect("serialize config");
    fs::write(config_path, yaml).expect("write config");
    world.wiki_directory = Some(value);
}

#[given(expr = "the Kanbus configuration has wiki_directory set to {string}")]
fn given_config_wiki_directory(world: &mut KanbusWorld, value: String) {
    let cwd = world.working_directory.as_ref().expect("working dir");
    let config_path = cwd.join(".kanbus.yml");
    let contents = fs::read_to_string(&config_path).expect("read config");
    let mut mapping: Mapping = serde_yaml::from_str(&contents).expect("parse config");
    mapping.insert(
        Value::String("wiki_directory".to_string()),
        Value::String(value),
    );
    let yaml = serde_yaml::to_string(&mapping).expect("serialize config");
    fs::write(config_path, yaml).expect("write config");
}

#[then(expr = "the wiki root should be {string}")]
fn then_wiki_root_is(world: &mut KanbusWorld, expected: String) {
    let cwd = world.working_directory.as_ref().expect("working dir");
    let wiki_path = cwd.join(&expected);
    assert!(
        wiki_path.exists(),
        "expected wiki root {expected} at {wiki_path:?}"
    );
}
